### PR TITLE
feat(node): add pnpm support

### PR DESCRIPTION
**Pull request by kecrily**
Check the neovim package installed with pnpm

### CODE REVIEW

File: runtime/autoload/health/provider.vim [hunk 1]
```diff
@@ -634,15 +634,24 @@ function! s:check_node() abort
...
           \  'Run in shell (if you use yarn): yarn global add neovim',
+          \  'Run in shell (if you use pnpm): pnpm install -g neovim',
           \  'You may disable this provider (and warning) by adding `let g:loaded_node_provider = 0` to your init.vim'])
     return
   endif
   call health#report_info('Nvim node.js host: '. host)
 
-  let manager = executable('npm') ? 'npm' : 'yarn'
+  if executable('npm')
+    " pwsh is the future
```
Comment (zeertzjq) [inline runtime/autoload/health/provider.vim:647]: What is this comment for?

Comment (kecrily) [reply runtime/autoload/health/provider.vim:647]: My fault, I'm not very familiar with the syntax of vim so I refer to other files.

File: runtime/autoload/health/provider.vim [hunk 2]
```diff
@@ -615,10 +615,10 @@ function! s:check_node() abort
     return
   endif
 
-  if !executable('node') || (!executable('npm') && !executable('yarn'))
+  if !executable('node') || (!executable('npm') && !executable('yarn') && !executable('pnpm'))
     call health#report_warn(
-          \ '`node` and `npm` (or `yarn`) must be in $PATH.',
-          \ ['Install Node.js and verify that `node` and `npm` (or `yarn`) commands work.'])
+          \ '`node` and `npm` (or `yarn`, `pnpm`) must be in $PATH.',
+          \ ['Install Node.js and verify that `node` and `npm` (or `yarn, `pnpm`) commands work.'])
```
Comment (justinmk) [inline runtime/autoload/health/provider.vim:621]:
```suggestion
          \ ['Install Node.js and verify that `node` and `npm` (or `yarn`, `pnpm`) commands work.'])
```

File: runtime/autoload/health/provider.vim [hunk 3]
```diff
@@ -634,15 +634,23 @@ function! s:check_node() abort
...
   call health#report_info('Nvim node.js host: '. host)
 
-  let manager = executable('npm') ? 'npm' : 'yarn'
+  if executable('npm')
+    let manager = 'npm'
+  elseif executable('yarn')
+    let manager = 'yarn'
+  elseif executable('pnpm')
+    let manager = 'pnpm'
+  endif
```
Comment (justinmk) [inline runtime/autoload/health/provider.vim:652]: the old logic defaulted to `yarn`, now `manager will be undefined if none are found, which will result in a weird error for the user.

**Review by justinmk — COMMENTED** [review]
This is fine as a stop-gap, but why are we using npm/yarn/etc at all ? Can we just try something like `node require('neovim')` and if that fails then report as appropriate?

File: runtime/autoload/health/provider.vim (hunk 3 above)
Comment (kecrily) [reply runtime/autoload/health/provider.vim:652]: Maybe we should default manager to npm and then change it if we find it in the other two managers.

**Comment by kecrily** [comment]
I don't think `require('neovim')` will work. Nodejs can only use packages directly under `node_modules` or the built-in packages.

It doesn't know where `neovim` is if we don't tell it the path to the package. And each package manager has a different global path, which ultimately brings us back to what we're doing now.

In deno we can use `https` links to load modules, but it not for node.

**Comment by justinmk** [comment]
Damn, I think the Windows CI failure is related to this PR:

```
ERROR    1 error, listed below:
ERROR    test/functional\plugin\health_spec.lua @ 35: :checkhealth completions can be listed via getcompletion()
test\functional\helpers.lua:106: EOF was received from Nvim. Likely the Nvim process crashed.

stack traceback:
	test\functional\helpers.lua:106: in function 'getcompletion'
	test/functional\plugin\health_spec.lua:37: in function <test/functional\plugin\health_spec.lua:35>

**Comment by zeertzjq** [comment]
> 
> ```
> ERROR    1 error, listed below:
> ERROR    test/functional\plugin\health_spec.lua @ 35: :checkhealth completions can be listed via getcompletion()
> test\functional\helpers.lua:106: EOF was received from Nvim. Likely the Nvim process crashed.
> 
> stack traceback:
> 	test\functional\helpers.lua:106: in function 'getcompletion'
> 	test/functional\plugin\health_spec.lua:37: in function <test/functional\plugin\health_spec.lua:35>
> ```

#19596

**Pull request merged**